components/pagination/pagination.ts

```diff
@@ -0,0 +1,166 @@
+    private _paginationClasses:boolean;
+
+    // Public members
+    public pageCount:number;
+
+    @Output()
+    public pageChange:EventEmitter<number>;
+
+    // Private members
+    private _maxSize?:number;
+    private _collectionSize:number;
+    private _pageSize:number;
+    private _page:number;
+    private _hasBoundaryLinks:boolean;
+    private _hasNavigation:boolean;
+
+    @Input()
+    public get maxSize():number {
+        return this._maxSize;
+    }
+
+    public set maxSize(value:number) {
+        this._maxSize = (value !== undefined) ? Math.max(value, 1) : undefined;
+    }
+
+    @Input()
+    public get collectionSize():number {
+        return this._collectionSize;
+    }
+
+    public set collectionSize(value:number) {
+        this._collectionSize = Math.max(value, 0);
+    }
+
+    @Input()
+    public get pageSize():number {
+        return this._pageSize;
+    }
+
+    public set pageSize(value:number) {
+        this._pageSize = value;
+    }
+
+    @Input()
+    public get hasNavigation():boolean {
+        return this._hasNavigation || this._maxSize < this.pageCount;
+    }
+
+    public set hasNavigation(value:boolean) {
+        this._hasNavigation = value;
+    }
+
+    @Input()
+    public get hasBoundaryLinks():boolean {
+        return this._hasBoundaryLinks;
+    }
+
+    public set hasBoundaryLinks(value:boolean) {
+        this._hasBoundaryLinks = value;
+    }
+
+    @Input()
+    public get page():number {
+        return this._page;
+    }
+
+    public set page(value:number) {
+        this.setPage(value);
+    }
+
+    public get pages():number[] {
+        const [start, end] = this.applyPagination();
+
+        return Array<number>(end - start)
+            .fill(start + 1)
+            .map((s, i) => s + i);
+    }
+
+    constructor() {
+        this._paginationClasses = true;
+        this.pageChange = new EventEmitter<number>();
+
+        this.collectionSize = 100;
+        this.pageSize = 10;
+        this.page = 1;
+        this.hasNavigation = false;
+        this.hasBoundaryLinks = false;
+    }
+
+    // Public methods
+    public hasPrevious():boolean {
+        return this.page > 1;
+    }
+
+    public hasNext():boolean {
+        return this.page < this.pageCount;
+    }
+
+    public setPage(newPage:number):void {
+        const value:number = (Number.isInteger(newPage)) ? Math.min(Math.max(newPage, 1), this.pageCount) : 1;
+        if (value !== this._page) {
+            this._page = value;
+            this.pageChange.emit(this._page);
+        }
+    }
+
+    // Lifecycle hooks
+    public ngOnChanges(changes:SimpleChanges):void {
+        console.log(changes);
+        this.updatePages(this.page);
+    }
+
+    // Private methods
+    private updatePages(newPage:number):void {
+        this.pageCount = Math.max(1, Math.ceil(this._collectionSize / this._pageSize));
+
+        this.setPage(newPage);
+    }
+
+    private applyPagination():[number, number] {
```

Comment:
This can be simplified by creating a local const `maxSize` defined as `this.maxSize || this.pageCount` - then you can delete the `if`, and the initial definitions of `start` & `end`.

```typescript
const maxSize = this.maxSize || this.pageCount;

const page = Math.ceil(this.page / maxSize) - 1;
let start = page * maxSize;
...
```